File: src/src/main/resources/mmlspark/RankingAdapterModel.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class RankingAdapterModel(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        itemCol (str): Column of items
        k (int): number of items (default: 10)
        labelCol (str): The name of the label column (default: label)
        minRatingsPerItem (int): min ratings for items > 0 (default: 1)
        minRatingsPerUser (int): min ratings for users > 0 (default: 1)
        mode (str): recommendation mode (default: allUsers)
        ratingCol (str): Column of ratings
        recommender (object): estimator for selection
        recommenderModel (object): recommenderModel
        userCol (str): Column of users
    """

    @keyword_only
    def __init__(self, itemCol=None, k=10, labelCol="label", minRatingsPerItem=1, minRatingsPerUser=1, mode="allUsers", ratingCol=None, recommender=None, recommenderModel=None, userCol=None):
        super(RankingAdapterModel, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.RankingAdapterModel")
        self._cache = {}
        self.itemCol = Param(self, "itemCol", "itemCol: Column of items")
        self.k = Param(self, "k", "k: number of items (default: 10)")
        self._setDefault(k=10)
        self.labelCol = Param(self, "labelCol", "labelCol: The name of the label column (default: label)")
        self._setDefault(labelCol="label")
        self.minRatingsPerItem = Param(self, "minRatingsPerItem", "minRatingsPerItem: min ratings for items > 0 (default: 1)")
        self._setDefault(minRatingsPerItem=1)
        self.minRatingsPerUser = Param(self, "minRatingsPerUser", "minRatingsPerUser: min ratings for users > 0 (default: 1)")
        self._setDefault(minRatingsPerUser=1)
        self.mode = Param(self, "mode", "mode: recommendation mode (default: allUsers)")
        self._setDefault(mode="allUsers")
        self.ratingCol = Param(self, "ratingCol", "ratingCol: Column of ratings")
        self.recommender = Param(self, "recommender", "recommender: estimator for selection", generateTypeConverter("recommender", self._cache, complexTypeConverter))
        self.recommenderModel = Param(self, "recommenderModel", "recommenderModel: recommenderModel", generateTypeConverter("recommenderModel", self._cache, complexTypeConverter))
        self.userCol = Param(self, "userCol", "userCol: Column of users")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, itemCol=None, k=10, labelCol="label", minRatingsPerItem=1, minRatingsPerUser=1, mode="allUsers", ratingCol=None, recommender=None, recommenderModel=None, userCol=None):
        """
        Set the (keyword only) parameters

        Args:

            itemCol (str): Column of items
            k (int): number of items (default: 10)
            labelCol (str): The name of the label column (default: label)
            minRatingsPerItem (int): min ratings for items > 0 (default: 1)
            minRatingsPerUser (int): min ratings for users > 0 (default: 1)
            mode (str): recommendation mode (default: allUsers)
            ratingCol (str): Column of ratings
            recommender (object): estimator for selection
            recommenderModel (object): recommenderModel
            userCol (str): Column of users
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setItemCol(self, value):
        """

        Args:

            itemCol (str): Column of items

        """
        self._set(itemCol=value)
        return self


    def getItemCol(self):
        """

        Returns:

            str: Column of items
        """
        return self.getOrDefault(self.itemCol)


    def setK(self, value):
        """

        Args:

            k (int): number of items (default: 10)

        """
        self._set(k=value)
        return self


    def getK(self):
        """

        Returns:

            int: number of items (default: 10)
        """
        return self.getOrDefault(self.k)


    def setLabelCol(self, value):
        """

        Args:

            labelCol (str): The name of the label column (default: label)

        """
        self._set(labelCol=value)
        return self


    def getLabelCol(self):
        """

        Returns:

            str: The name of the label column (default: label)
        """
        return self.getOrDefault(self.labelCol)


    def setMinRatingsPerItem(self, value):
        """

        Args:

            minRatingsPerItem (int): min ratings for items > 0 (default: 1)

        """
        self._set(minRatingsPerItem=value)
        return self


    def getMinRatingsPerItem(self):
        """

        Returns:

            int: min ratings for items > 0 (default: 1)
        """
        return self.getOrDefault(self.minRatingsPerItem)


    def setMinRatingsPerUser(self, value):
        """

        Args:

            minRatingsPerUser (int): min ratings for users > 0 (default: 1)

        """
        self._set(minRatingsPerUser=value)
        return self


    def getMinRatingsPerUser(self):
        """

        Returns:

            int: min ratings for users > 0 (default: 1)
        """
        return self.getOrDefault(self.minRatingsPerUser)


    def setMode(self, value):
        """

        Args:

            mode (str): recommendation mode (default: allUsers)

        """
        self._set(mode=value)
        return self


    def getMode(self):
        """

        Returns:

            str: recommendation mode (default: allUsers)
        """
        return self.getOrDefault(self.mode)


    def setRatingCol(self, value):
        """

        Args:

            ratingCol (str): Column of ratings

        """
        self._set(ratingCol=value)
        return self


    def getRatingCol(self):
        """

        Returns:

            str: Column of ratings
        """
        return self.getOrDefault(self.ratingCol)


    def setRecommender(self, value):
        """

        Args:

            recommender (object): estimator for selection

        """
        self._set(recommender=value)
        return self


    def getRecommender(self):
        """

        Returns:

            object: estimator for selection
        """
        return self._cache.get("recommender", None)


    def setRecommenderModel(self, value):
        """

        Args:

            recommenderModel (object): recommenderModel

        """
        self._set(recommenderModel=value)
        return self


    def getRecommenderModel(self):
        """

        Returns:

            object: recommenderModel
        """
        return self._cache.get("recommenderModel", None)


    def setUserCol(self, value):
        """

        Args:

            userCol (str): Column of users

        """
        self._set(userCol=value)
        return self


    def getUserCol(self):
        """

        Returns:

            str: Column of users
        """
        return self.getOrDefault(self.userCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.RankingAdapterModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=RankingAdapterModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".RankingAdapterModel"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/RankingAdapterModel.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Utils;

using SynapseML.Dotnet.Utils;


namespace Synapse.ML.Recommendation
{
    /// <summary>
    /// <see cref="RankingAdapterModel"/> implements RankingAdapterModel
    /// </summary>
    public class RankingAdapterModel : JavaModel<RankingAdapterModel>, IJavaMLWritable, IJavaMLReadable<RankingAdapterModel>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.recommendation.RankingAdapterModel";

        /// <summary>
        /// Creates a <see cref="RankingAdapterModel"/> without any parameters.
        /// </summary>
        public RankingAdapterModel() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="RankingAdapterModel"/> with a UID that is used to give the
        /// <see cref="RankingAdapterModel"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public RankingAdapterModel(string uid) : base(s_className, uid)
        {
        }

        internal RankingAdapterModel(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets itemCol value for <see cref="itemCol"/>
        /// </summary>
        /// <param name="itemCol">
        /// Column of items
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetItemCol(string value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setItemCol", (object)value));
        
        /// <summary>
        /// Sets k value for <see cref="k"/>
        /// </summary>
        /// <param name="k">
        /// number of items
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetK(int value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setK", (object)value));
        
        /// <summary>
        /// Sets labelCol value for <see cref="labelCol"/>
        /// </summary>
        /// <param name="labelCol">
        /// The name of the label column
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetLabelCol(string value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setLabelCol", (object)value));
        
        /// <summary>
        /// Sets minRatingsPerItem value for <see cref="minRatingsPerItem"/>
        /// </summary>
        /// <param name="minRatingsPerItem">
        /// min ratings for items > 0
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetMinRatingsPerItem(int value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setMinRatingsPerItem", (object)value));
        
        /// <summary>
        /// Sets minRatingsPerUser value for <see cref="minRatingsPerUser"/>
        /// </summary>
        /// <param name="minRatingsPerUser">
        /// min ratings for users > 0
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetMinRatingsPerUser(int value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setMinRatingsPerUser", (object)value));
        
        /// <summary>
        /// Sets mode value for <see cref="mode"/>
        /// </summary>
        /// <param name="mode">
        /// recommendation mode
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetMode(string value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setMode", (object)value));
        
        /// <summary>
        /// Sets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <param name="ratingCol">
        /// Column of ratings
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetRatingCol(string value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setRatingCol", (object)value));
        
        /// <summary>
        /// Sets recommender value for <see cref="recommender"/>
        /// </summary>
        /// <param name="recommender">
        /// estimator for selection
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetRecommender<M>(JavaEstimator<M> value) where M : JavaModel<M> =>
            WrapAsRankingAdapterModel(Reference.Invoke("setRecommender", (object)value));
        
        /// <summary>
        /// Sets recommenderModel value for <see cref="recommenderModel"/>
        /// </summary>
        /// <param name="recommenderModel">
        /// recommenderModel
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetRecommenderModel(JavaTransformer value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setRecommenderModel", (object)value));
        
        /// <summary>
        /// Sets userCol value for <see cref="userCol"/>
        /// </summary>
        /// <param name="userCol">
        /// Column of users
        /// </param>
        /// <returns> New RankingAdapterModel object </returns>
        public RankingAdapterModel SetUserCol(string value) =>
            WrapAsRankingAdapterModel(Reference.Invoke("setUserCol", (object)value));

        
        /// <summary>
        /// Gets itemCol value for <see cref="itemCol"/>
        /// </summary>
        /// <returns>
        /// itemCol: Column of items
        /// </returns>
        public string GetItemCol() =>
            (string)Reference.Invoke("getItemCol");
        
        
        /// <summary>
        /// Gets k value for <see cref="k"/>
        /// </summary>
        /// <returns>
        /// k: number of items
        /// </returns>
        public int GetK() =>
            (int)Reference.Invoke("getK");
        
        
        /// <summary>
        /// Gets labelCol value for <see cref="labelCol"/>
        /// </summary>
        /// <returns>
        /// labelCol: The name of the label column
        /// </returns>
        public string GetLabelCol() =>
            (string)Reference.Invoke("getLabelCol");
        
        
        /// <summary>
        /// Gets minRatingsPerItem value for <see cref="minRatingsPerItem"/>
        /// </summary>
        /// <returns>
        /// minRatingsPerItem: min ratings for items > 0
        /// </returns>
        public int GetMinRatingsPerItem() =>
            (int)Reference.Invoke("getMinRatingsPerItem");
        
        
        /// <summary>
        /// Gets minRatingsPerUser value for <see cref="minRatingsPerUser"/>
        /// </summary>
        /// <returns>
        /// minRatingsPerUser: min ratings for users > 0
        /// </returns>
        public int GetMinRatingsPerUser() =>
            (int)Reference.Invoke("getMinRatingsPerUser");
        
        
        /// <summary>
        /// Gets mode value for <see cref="mode"/>
        /// </summary>
        /// <returns>
        /// mode: recommendation mode
        /// </returns>
        public string GetMode() =>
            (string)Reference.Invoke("getMode");
        
        
        /// <summary>
        /// Gets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <returns>
        /// ratingCol: Column of ratings
        /// </returns>
        public string GetRatingCol() =>
            (string)Reference.Invoke("getRatingCol");
        
        
        /// <summary>
        /// Gets recommender value for <see cref="recommender"/>
        /// </summary>
        /// <returns>
        /// recommender: estimator for selection
        /// </returns>
        public IEstimator<object> GetRecommender()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getRecommender");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaPipelineStage),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out IEstimator<object> instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets recommenderModel value for <see cref="recommenderModel"/>
        /// </summary>
        /// <returns>
        /// recommenderModel: recommenderModel
        /// </returns>
        public JavaTransformer GetRecommenderModel()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getRecommenderModel");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets userCol value for <see cref="userCol"/>
        /// </summary>
        /// <returns>
        /// userCol: Column of users
        /// </returns>
        public string GetUserCol() =>
            (string)Reference.Invoke("getUserCol");

        
        /// <summary>
        /// Loads the <see cref="RankingAdapterModel"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="RankingAdapterModel"/> was saved to</param>
        /// <returns>New <see cref="RankingAdapterModel"/> object, loaded from path.</returns>
        public static RankingAdapterModel Load(string path) => WrapAsRankingAdapterModel(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<RankingAdapterModel> Read() =>
            new JavaMLReader<RankingAdapterModel>((JvmObjectReference)Reference.Invoke("read"));

        private static RankingAdapterModel WrapAsRankingAdapterModel(object obj) =>
            new RankingAdapterModel((JvmObjectReference)obj);

        
    }
}

        